src/main/java/de/tum/cit/aet/artemis/programming/service/localvc/LocalVCServletService.java

```diff
@@ -278,18 +287,66 @@ private void savePreliminaryVcsAccessLogForHTTPs(HttpServletRequest request, Loc
             var ipAddress = request.getRemoteAddr();
             var authenticationMechanism = resolveHTTPSAuthenticationMechanism(request.getHeader(LocalVCServletService.AUTHORIZATION_HEADER), user);
 
-            String commitHash = null;
+            String finalCommitHash = getCommitHash(localVCRepositoryUri);
+            RepositoryActionType finalRepositoryAction = repositoryAction == RepositoryActionType.WRITE ? RepositoryActionType.PUSH : RepositoryActionType.PULL;
+            vcsAccessLogService.ifPresent(service -> service.saveAccessLog(user, participation, finalRepositoryAction, authenticationMechanism, finalCommitHash, ipAddress));
+        }
+    }
+
+    /**
+     * Logs a failed attempt to access a repository.
+     *
+     * @param request                  An optional {@link HttpServletRequest} containing request details if access was attempted via HTTPS.
+     * @param session                  An optional {@link ServerSession} containing session details if access was attempted via SSH.
+     * @param repositoryTypeOrUserName A string representing either the repository type or the username associated with the repository.
+     * @param exercise                 The {@link Exercise} associated with the repository.
+     * @param localVCRepositoryUri     The {@link LocalVCRepositoryUri} representing the repository location.
+     * @param user                     The {@link User} attempting the access.
+     * @param repositoryAction         The {@link RepositoryActionType} action that was attempted.
+     */
+    public void saveFailedAccessVcsAccessLog(Optional<HttpServletRequest> request, Optional<ServerSession> session, String repositoryTypeOrUserName, Exercise exercise,
```

Comment:
This method somehow seems like it's doing too much. Using some if-else for different use cases makes it kind of hard to understand and keep context in mind.

I think it's still fine, but one _could_ consider splitting it up. Fine with it, tho

Reply:
I agree here tbh, I refactored it a bit and extracted the ifs